_posts/2023-01-21-pthread-cancellation.markdown

```diff
@@ -0,0 +1,168 @@
+
+As an example, let's investigate ``accept()``. Suppose there is a thread which listens for connections, and is blocking inside an ``accept()`` call. Now it's cancelled. What should happen? If mlibc unconditionally peformed a cancellation request, then this would cause resource leaks. ``accept()`` can allocate a file descriptor. If a file descriptor gets allocated, and mlibc cancels the thread, then the file descriptor is lost.
+
+What _should_ happen is mlibc determines whether or not the syscall produced _side-effects_. If the syscall produced side-effects, mlibc should defer the cancellation request, and let the thread clean up its side-effects first. However, if no side-effects were caused, then the thread can be canceled. A mechanism for this will be described in the following section.
+
+### Cancelling
+
+As said before, to cancel a thread ``pthread_cancel`` can be used. The logic for this function is quite simple. The cancel trigger bit in the TCB is set atomically, and if cancellation is enabled, a SIGCANCEL is delivered to the target thread.
+
+The cancel logic is performed by ``sigcancel_handler()``, the signal handler for SIGCANCEL. A [static constructor](https://github.com/managarm/mlibc/blob/a6cbfbaca35fbe24a56aae720ae67eb45925d8ce/options/posix/generic/pthread-stubs.cpp#L260-L273) is used to install the signal handler at startup. A summarized version of the function is shown below.
+
+```cpp
+ucontext_t *uctx = static_cast<ucontext_t*>(ucontext);
+// The function could be called from other signals, or from another
+// process, in which case we should do nothing.
+if (signal != SIGCANCEL || info->si_pid != getpid() ||
+            info->si_code != SI_TKILL)
+    return;
+
+auto tcb = reinterpret_cast<Tcb*>(mlibc::get_current_tcb());
+int old_value = tcb->cancelBits;
+
+if (!(old_value & tcbCancelAsyncBit) &&
+            mlibc::sys_before_cancellable_syscall && !mlibc::sys_before_cancellable_syscall(uctx))
+        return;
+
+// ...
+// Set cancel trigger and canceling bit atomically,
+// then call __mlibc_do_cancel()
+//
+```
+
+There are two notable conditions in this function. First the function checks if the cancellation actually came from the current process, else other programs could theoretically shoot down arbitrary threads.
+
+Secondly, the function checks if deferred cancellation is enabled. This is to handle the issue described in the previous section. First it checks if the ``sys_before_cancellable_syscall`` sysdep [^1] is available. If it is, then it calls that with the thread's ``ucontext``, which is a struct containing the current register state of the thread.
+
+On the linux sysdeps, ``sys_before_cancellable_syscall`` is implemented as follows (code can be found [here](https://github.com/managarm/mlibc/blob/f2c6b8cd7fe67b89d12b23ba0a82e174ba2b7c0a/sysdeps/linux/generic/sysdeps.cpp#L365-L370)).
+
+```cpp
+int sys_before_cancellable_syscall(ucontext_t *uct) {
+	auto pc = reinterpret_cast<void*>(uct->uc_mcontext.rip);
```

Comment:
Not super relevant to this post, but you could implement this in a cleaner way:
```cpp
auto pc = uct->uc_mcontext.rip; // keep as uint64_t (comparing pointers is dodgy from the standard's POV)
return pc >= __mlibc_syscall_begin && pc <= __mlibc_syscall_end;

Reply:
Yeah true, might do that in upstream mlibc sometime.

Reply:
Well, you could fix it here too 😄 makes it a little easier to follow